uuid/v7.ts

```diff
@@ -0,0 +1,135 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
+/**
+ * Functions for working with UUID Version 7 strings.
+ *
+ * UUID Version 7 is defined in {@link https://www.rfc-editor.org/rfc/rfc9562.html#section-5.7 | RFC 9562}.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ *
+ * @example
+ * ```ts
+ * import { generate, validate, extractTimestamp } from "@std/uuid/v7";
+ * import { assert, assertEquals } from "@std/assert";
+ *
+ * const uuid = generate();
+ * assert(validate(uuid));
+ * assertEquals(extractTimestamp("017f22e2-79b0-7cc3-98c4-dc0c0c07398f"), 1645557742000);
+ * ```
+ *
+ * @module
+ */
+
+import { bytesToUuid } from "./_common.ts";
+
+const UUID_RE =
+  /^[0-9a-f]{8}-[0-9a-f]{4}-[7][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
+/**
+ * Determines whether a string is a valid
+ * {@link https://www.rfc-editor.org/rfc/rfc9562.html#section-5.7 | UUIDv7}.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ *
+ * @param id UUID value.
+ *
+ * @returns `true` if the string is a valid UUIDv7, otherwise `false`.
+ *
+ * @example Usage
+ * ```ts
+ * import { validate } from "@std/uuid/v7";
+ * import { assert, assertFalse } from "@std/assert";
+ *
+ * assert(validate("017f22e2-79b0-7cc3-98c4-dc0c0c07398f"));
+ * assertFalse(validate("fac8c1e0-ad1a-4204-a0d0-8126ae84495d"));
+ * ```
+ */
+export function validate(id: string): boolean {
+  return UUID_RE.test(id);
+}
+
+/**
+ * Options for {@linkcode generate}.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ */
+export interface GenerateOptions {
+  /**
+   * Unix epoch timestamp in milliseconds.
+   *
+   * @default {Date.now()}
+   */
+  timestamp?: number;
+}
+
+/**
+ * Generates a {@link https://www.rfc-editor.org/rfc/rfc9562.html#section-5.7 | UUIDv7}.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ *
+ * @throws {RangeError} If the timestamp is not a non-negative integer.
+ *
+ * @param options Options for generating the UUIDv7.
+ * @returns Returns a UUIDv7 string
+ *
+ * @example Usage
+ * ```ts
+ * import { generate, validate } from "@std/uuid/v7";
+ * import { assert } from "@std/assert";
+ *
+ * const uuid = generate();
+ * assert(validate(uuid));
+ * ```
+ */
+export function generate(options: GenerateOptions = {}): string {
```

Comment:
I think it's safe to do this. WDYT?
```suggestion
export function generate(timestamp = Date.now()): string {
```

Reply:
Keeping GenerateOptions allows adding the optional features for monotonicity approach be chosen  in the future. If that is desired, if monotonicity is not an option and is always the same, having just timestamp if works.
 
e.g something like the following
```ts
generate({ monotonicity: "counter" })
generate({ nano: true, counter: true })
```
[optional features](https://www.rfc-editor.org/rfc/rfc9562.html#section-5.7-3.1)

Reply:
No consistency issues either way
- v1 `generate(options: GenerateOptions = {}): string`
- v3 `generate(namespace: string, data: Uint8Array): Promise<string>`
- v4 `crypto.randomUUID()`
- v5 `generate(namespace: string, data: Uint8Array): Promise<string>`

Reply:
It's possible we'll add more options. It's also possible we won't. This is an experimental API. If we add more options, we can introduce a `GenerateOptions`. I think let's just go with a single `timestamp = Date.now()` argument for now 🙂

Reply:
Sweet, I will update to a use a single timestamp argument then

Reply:
Done [here](https://github.com/denoland/std/pull/5887/commits/7dfa002d2fdd3dd9faac2ce4a9ee37d597a58d4b).

As a side effect of this change, even when the default `Date.now()` is used, the timestamp is still validated to ensure it’s a non-negative integer. 
This differs from the previous behaviour, where validation was skipped if the default `Date.now()` was used.